### PR TITLE
Add Guided filter

### LINKED ISSUE #2241: Add Guided filter
## 🚀 Feature

Add Guided filter. It is an edge-preserving image smoothing filter, similar to bilateral filter.

https://kaiminghe.github.io/publications/eccv10guidedfilter.pdf

https://en.wikipedia.org/wiki/Guided_filter

OpenCV:

- https://docs.opencv.org/4.x/da/d17/group__ximgproc__filters.html#ga86813d59f8452a9600e979c6287805f5
- https://github.com/opencv/opencv_contrib/blob/4.x/modules/ximgproc/src/guided_filter.cpp

Matlab: https://www.mathworks.com/help/images/ref/imguidedfilter.html

Some PyTorch implementations

- https://github.com/wuhuikai/DeepGuidedFilter/blob/master/GuidedFilteringLayer/GuidedFilter_PyTorch/guided_filter_pytorch/guided_filter.py (grayscale image only)
- https://github.com/perrying/guided-filter-pytorch/blob/master/guided_filter.py (grayscale and RGB)

Guided filter for 1-channel (grayscale) image is simple and straight-forward. However, for multi-channel (RGB), we need to calculate the covariance matrix for the channels (treat each channel as a random variable), and calculate matrix inverse. perrying's implementation hard-codes this calculation.

## Pitch

Perhaps we can add guided filter for single channel first? Then we can decide if we want to only support RGB images (then it's possible to hard-code the calculations) or >3-channel images.

## Additional context

______________________________________________________________________

#### Consider also to contribute to Kornia universe projects :)

<sub>

- [**Tutorials**](https://github.com/kornia/tutorials): our repository containing the tutorials.

</sub>

### PR DESCRIPTION
**Pull request by gau-nernst**
#### Changes

Fixes #2241 

This also implements Fast Guided filter (https://arxiv.org/abs/1505.00996), where guidance and input are downsampled before computing a and b.

#### Type of change

- [ ] 📚  Documentation Update
- [ ] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📝 This change requires a documentation update

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

**Comment by gau-nernst** [comment]
@edgarriba Can you take a look and give me some feedback before I finalize other details (add docs, tests, example image...)?

Currently I keep 2 implementations: (1) for single-channel (grayscale) guidance, and (2) for multi-channel (e.g. RGB) guidance. The single-channel implementation should be faster and easier to read. In the multi-channel implementation, since the matrix ops like inverse or matmul require the matrix dims to be the last dimensions, I need quite a lot of transpose ops.

I tested against OpenCV implementation and the difference was in the order of 1e-5, for both single channel and multi channel. Note that OpenCV uses hard-coded BORDER_REFLECT mode ([here](https://github.com/opencv/opencv_contrib/blob/853144ef93c4ffa55661619b861539090943c5b6/modules/ximgproc/src/guided_filter.cpp#L162), which also reflects the outermost values (BORDER_REFLECT_101 will match the behavior of PyTorch's "reflect" padding).

**Comment by gau-nernst** [comment]
image


```python
import io

import cv2
import matplotlib.pyplot as plt
import numpy as np
import requests
import torch
from PIL import Image

from kornia.filters import guided_blur


img = io.BytesIO(requests.get("https://upload.wikimedia.org/wikipedia/en/7/7d/Lenna_%28test_image%29.png").content)
x_np = np.array(Image.open(img))
x_np = x_np.astype(np.float32) / 255
x = torch.from_numpy(x_np).permute(2, 0, 1).unsqueeze(0)

k = 11
eps = 0.1
out1 = guided_blur(x, x, k, eps, border_type="replicate")
out2 = cv2.ximgproc.guidedFilter(x_np, x_np, (k-1)//2, eps)

diff = np.abs(out1.squeeze(0).permute(1, 2, 0).numpy().squeeze() - out2)

fig, axes = plt.subplots(1, 3, figsize=(30,10))

axes[0].imshow(out1.squeeze(0).permute(1, 2, 0).numpy(), vmin=0, vmax=1)
axes[0].set_title("pytorch")

axes[1].imshow(out2, vmin=0, vmax=1)
axes[1].set_title("opencv")

axes[2].imshow(diff, vmin=0, vmax=1)
axes[2].set_title("diff")

plt.show()
```

File: kornia/filters/guided.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+
+    def __init__(
+        self, kernel_size: tuple[int, int] | int, eps: float | Tensor, border_type: str = 'reflect', subsample: int = 1
+    ):
+        super().__init__()
+        self.kernel_size = kernel_size
+        if isinstance(eps, Tensor):
+            self.register_buffer("eps", eps)
+        else:
+            self.eps = eps
```
Comment (johnnv1) [inline kornia/filters/guided.py:203]:
```suggestion
        self.register_buffer("eps", as_tensor(eps))
```

File: kornia/filters/guided.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    if input is not guidance:
+        KORNIA_CHECK_IS_TENSOR(input)
+        KORNIA_CHECK_SHAPE(input, ['B', 'C', 'H', 'W'])
+        KORNIA_CHECK(
+            (guidance.shape[0] == input.shape[0]) and (guidance.shape[-2:] == input.shape[-2:]),
+            "guidance and input should have the same batch size and spatial dimensions",
+        )
+
+    kernel_size = _unpack_2d_ks(kernel_size)
+    if guidance.shape[1] == 1:
```
Comment (johnnv1) [inline kornia/filters/guided.py:159]: we should add raise as optional in KORNIA_CHECK so we can use it like:

```python
if KORNIA_CHECK_IS_GRAY(guidance, raise=False):
    ...
```
cc\ @edgarriba

File: kornia/filters/guided.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+    KORNIA_CHECK_IS_TENSOR(guidance)
+    KORNIA_CHECK_SHAPE(guidance, ['B', 'C', 'H', 'W'])
+    if input is not guidance:
+        KORNIA_CHECK_IS_TENSOR(input)
+        KORNIA_CHECK_SHAPE(input, ['B', 'C', 'H', 'W'])
+        KORNIA_CHECK(
+            (guidance.shape[0] == input.shape[0]) and (guidance.shape[-2:] == input.shape[-2:]),
+            "guidance and input should have the same batch size and spatial dimensions",
+        )
```
Comment (johnnv1) [inline kornia/filters/guided.py:156]: I have seen it in some places, maybe we should add an option on `KORNIA_CHECK_SAME_SHAPE` to choose the dims. cc\ @edgarriba

**Review by johnnv1 — APPROVED** [review]
pretty nice patch 🚀

File: kornia/filters/guided.py (hunk 1 above)
Comment (gau-nernst) [reply kornia/filters/guided.py:203]: There is another case, when `eps` is trainable, we need to use `self.register_parameter()`. How should we handle it? I saw one approach here #1923, but it doesn't feel convenient to use, as the user needs to wrap `eps` in `nn.Parameter()`. How about having an argument `trainable_eps` in `__init__()`?

Comment (johnnv1) [reply kornia/filters/guided.py]: has any gain in the eps been a Tensor instead of a simple float here?

Comment (gau-nernst) [reply kornia/filters/guided.py]: From what I see, GaussianBlur also accepts sigma as Tensor so I follow it

https://github.com/kornia/kornia/blob/0ff08c72d5b5da101f1e57916cba892c7b3a2905/kornia/filters/gaussian.py#L104-L110

You are right, for non-trainable `eps`, there is no benefit to have it as a Tensor. I will make `eps` as `float` only. (Perhaps we should update other modules to not accept tensor also? like sigma in GaussianBlur)

What do you think about having `trainable_eps` as an argument in `__init__()`? To have `eps` as trainable parameter. (Then we will do something like `self.register_parameter("eps", nn.Parameter(torch.tensor([eps])))`)

File: kornia/filters/guided.py (hunk 3 above)
Comment (edgarriba) [reply kornia/filters/guided.py:156]: that's a good improvement to do

File: kornia/filters/guided.py (hunk 2 above)
Comment (edgarriba) [reply kornia/filters/guided.py:158]: i like it /cc @ducha-aiki

**Review by edgarriba — APPROVED** [review]
LGTM -- amazing work !!
